File: benchmarks/wrappers/cloudflare/python/container/storage.py
"""
Storage module for Cloudflare Python Containers
Uses a Container outbound handler to access the Worker R2 binding's R2 binding
"""
import io
import os
import json
import urllib.request
import urllib.parse


class storage:
    """R2 storage client for containers using a Worker outbound binding handler"""
    instance = None
    outbound_url = "http://sebs.r2"
    
    def __init__(self):
        # R2 calls use the Worker outbound handler virtual host.
        self.r2_enabled = True
    
    @staticmethod
    def init_instance(entry=None):
        """Initialize singleton instance"""
        if storage.instance is None:
            storage.instance = storage()
        return storage.instance
    
    @staticmethod
    def get_instance():
        """Get singleton instance"""
        if storage.instance is None:
            storage.init_instance()
        return storage.instance
    
    @staticmethod
    def unique_name(name):
        """Generate unique name for file"""
        import uuid
        name_part, extension = os.path.splitext(name)
        return f'{name_part}.{str(uuid.uuid4()).split("-")[0]}{extension}'
    
    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _post_json(self, url: str, body: bytes = b'', content_type: str = 'application/octet-stream'):
        """POST *body* to *url* and return the parsed JSON response."""
        req = urllib.request.Request(url, data=body, method='POST')
        req.add_header('Content-Type', content_type)
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read().decode('utf-8'))


    def _upload_bytes(self, key: str, data: bytes) -> str:
        return self._single_upload(key, data)

    def _single_upload(self, key: str, data: bytes) -> str:
        params = urllib.parse.urlencode({'key': key})
        url = f"{storage.outbound_url}/r2/upload?{params}"
        result = self._post_json(url, data)
        return result['key']


    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def upload_stream(self, bucket: str, key: str, data):
        """Upload data to R2 via the outbound handler"""
        if not self.r2_enabled:
            print("Warning: R2 not configured, skipping upload")
            return key
        
        if not storage.outbound_url:
            raise RuntimeError("Outbound handler URL is not configured - cannot access R2")
        
        # Handle BytesIO objects
        if isinstance(data, io.BytesIO):
            data = data.getvalue()
        
        # Convert to bytes if needed
        if isinstance(data, str):
            data = data.encode('utf-8')

        unique_key = self.unique_name(key)
        
        try:
            return self._upload_bytes(unique_key, data)
        except Exception as e:
            print(f"R2 upload error: {e}")
            raise RuntimeError(f"Failed to upload to R2: {e}")

    def download_stream(self, bucket: str, key: str) -> bytes:
        """Download data from R2 via the outbound handler"""
        if not self.r2_enabled:
            raise RuntimeError("R2 not configured")
        
        if not storage.outbound_url:
            raise RuntimeError("Outbound handler URL is not configured - cannot access R2")
        
        # Download through the Worker outbound handler.
        params = urllib.parse.urlencode({'bucket': bucket, 'key': key})
        url = f"{storage.outbound_url}/r2/download?{params}"
        
        try:
            with urllib.request.urlopen(url) as response:
                return response.read()
        except urllib.error.HTTPError as e:
            if e.code == 404:
                raise RuntimeError(f"Object not found: {key}")
            else:
                raise RuntimeError(f"Failed to download from R2: {e}")
        except Exception as e:
            print(f"R2 download error: {e}")
            raise RuntimeError(f"Failed to download from R2: {e}")
    
    def upload(self, bucket, key, filepath, unique_name=True):
        """Upload file from disk."""
        upload_key = self.unique_name(key) if unique_name else key
        with open(filepath, 'rb') as f:
            data = f.read()
        try:
            self._upload_bytes(upload_key, data)
        except Exception as e:
            raise RuntimeError(f"Failed to upload to R2: {e}")
        return upload_key
    
    def _upload_with_key(self, bucket: str, key: str, data):
        """Upload data to R2 via the outbound handler with exact key (internal method)"""
        if not self.r2_enabled:
            print("Warning: R2 not configured, skipping upload")
            return
        
        if not storage.outbound_url:
            raise RuntimeError("Outbound handler URL is not configured - cannot access R2")
        
        # Handle BytesIO objects
        if isinstance(data, io.BytesIO):
            data = data.getvalue()
        
        # Convert to bytes if needed
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        try:
            result_key = self._upload_bytes(key, data)
            print(f"[storage._upload_with_key] Upload successful, key={result_key}")
        except Exception as e:
            print(f"R2 upload error: {e}")
            raise RuntimeError(f"Failed to upload to R2: {e}")
    
    def download(self, bucket, key, filepath):
        """Download file to disk"""
        data = self.download_stream(bucket, key)
        os.makedirs(os.path.dirname(filepath), exist_ok=True)
        with open(filepath, 'wb') as f:
            f.write(data)

    def download_within_range(self, bucket: str, key: str, start_bytes: int, end_bytes: int) -> str:
        """Download a byte range of an object from R2 via the worker proxy."""
        if not self.r2_enabled:
            raise RuntimeError("R2 not configured")

        if not storage.worker_url:
            raise RuntimeError("Worker URL not set - cannot access R2")

        params = urllib.parse.urlencode({'bucket': bucket, 'key': key})
        url = f"{storage.worker_url}/r2/download?{params}"

        req = urllib.request.Request(url)
        req.add_header('Range', f'bytes={start_bytes}-{end_bytes}')

        try:
            with urllib.request.urlopen(req) as response:
                return response.read().decode('utf-8')
        except urllib.error.HTTPError as e:
            if e.code in (206, 200):
                return e.read().decode('utf-8')
            raise RuntimeError(f"Failed to download range from R2: {e}")
        except Exception as e:
            raise RuntimeError(f"Failed to download range from R2: {e}")

    def list_directory(self, bucket, prefix):
        """List all object keys with a given prefix."""
        if not storage.worker_url:
            raise RuntimeError("Worker URL not set - cannot access R2")
        params = urllib.parse.urlencode({'bucket': bucket, 'prefix': prefix})
        list_url = f"{storage.worker_url}/r2/list?{params}"
        with urllib.request.urlopen(list_url) as response:
            result = json.loads(response.read().decode('utf-8'))
            return [obj['key'] for obj in result.get('objects', [])]

    def download_directory(self, bucket, prefix, local_path):
        """
        Download all files with a given prefix to a local directory.
        Lists objects via /r2/list endpoint and downloads each one in parallel.
        """
        import concurrent.futures

        if not storage.outbound_url:
            raise RuntimeError("Outbound handler URL is not configured - cannot access R2")
        
        # Create local directory
        os.makedirs(local_path, exist_ok=True)
        
        # List objects with prefix through the Worker outbound handler.
        params = urllib.parse.urlencode({'bucket': bucket, 'prefix': prefix})
        list_url = f"{storage.outbound_url}/r2/list?{params}"
        
        try:
            with urllib.request.urlopen(list_url) as response:
                result = json.loads(response.read().decode('utf-8'))
                objects = result.get('objects', [])
                
                print(f"Found {len(objects)} objects with prefix '{prefix}'")

                def _download_one(obj):
                    obj_key = obj['key']
                    local_file_path = os.path.join(local_path, obj_key)
                    local_dir = os.path.dirname(local_file_path)
                    if local_dir:
                        os.makedirs(local_dir, exist_ok=True)
                    print(f"Downloading {obj_key} to {local_file_path}")
                    self.download(bucket, obj_key, local_file_path)

                # Download all objects in parallel (up to 16 concurrent)
                with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
                    futures = [executor.submit(_download_one, obj) for obj in objects]
                    for fut in concurrent.futures.as_completed(futures):
                        fut.result()  # re-raise any exception

                return local_path
                
        except Exception as e:
            print(f"Error listing/downloading directory: {e}")
            raise RuntimeError(f"Failed to download directory: {e}")

File: benchmarks/wrappers/cloudflare/nodejs/container/storage.js
const fs = require('fs');
const path = require('path');
const uuid = require('uuid');
const debug = require('util').debuglog('sebs');


/**
 * Storage module for Cloudflare Node.js Containers.
 *
 * Cloudflare documents the Container-to-binding integration path through a
 * Worker outbound handler and a virtual host (`http://sebs.r2`). The handler
 * runs in the Worker, where the R2 binding is available, and performs the
 * actual get/put/list calls.
 *
 * R2 also exposes a supported S3-compatible HTTPS API. A container can use
 * that alternative directly, but it requires provisioning and injecting R2
 * access keys into the container. It is separate from the documented
 * Container-to-binding path used by these benchmarks.
 *
 */

class storage {
  constructor() {
    this.r2_enabled = true;
  }
  
  static outbound_url = 'http://sebs.r2';

  static unique_name(name) {
    const parsed = path.parse(name);
    const uuid_name = uuid.v4().split('-')[0];
    return path.join(parsed.dir, `${parsed.name}.${uuid_name}${parsed.ext}`);
  }

  static init_instance(entry) {
    if (!storage.instance) {
      storage.instance = new storage();
    }
    return storage.instance;
  }
  
  
  static get_instance() {
    if (!storage.instance) {
      storage.init_instance();
    }
    return storage.instance;
  }

  _toBuffer(data) {
    if (Buffer.isBuffer(data)) {
      return data;
    }
    if (typeof data === 'string') {
      return Buffer.from(data, 'utf-8');
    }
    if (data instanceof ArrayBuffer) {
      return Buffer.from(data);
    }
    return Buffer.from(String(data), 'utf-8');
  }

  async _postJson(url, body = Buffer.alloc(0), contentType = null) {
    const options = {
      method: 'POST',
      body,
    };

    if (contentType) {
      options.headers = { 'Content-Type': contentType };
    }

    const response = await fetch(url, options);

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${await response.text()}`);
    }

    return response.json();
  }

  async _single_upload(key, buffer) {
    const params = new URLSearchParams({ key });
    const url = `${storage.outbound_url}/r2/upload?${params}`;
    const result = await this._postJson(url, buffer);
    return result.key;
  }



  async _upload_bytes(key, buffer) {
    return this._single_upload(key, buffer);
  }

  async upload_stream(bucket, key, data) {
    if (!this.r2_enabled) {
      debug('R2 not configured, skipping upload');
      return key;
    }


    const unique_key = storage.unique_name(key);
    const buffer = this._toBuffer(data);

    try {
      return await this._upload_bytes(unique_key, buffer);
    } catch (error) {
      debug('R2 upload error: %o', error);
      throw new Error(`Failed to upload to R2: ${error.message}`);
    }
  }

  async download_stream(bucket, key) {
    if (!this.r2_enabled) {
      throw new Error('R2 not configured');
    }


    // Download through the Worker outbound handler.
    const params = new URLSearchParams({ bucket, key });
    const url = `${storage.outbound_url}/r2/download?${params}`;

    try {
      const response = await fetch(url);

      if (response.status === 404) {
        throw new Error(`Object not found: ${key}`);
      }

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${await response.text()}`);
      }

      const arrayBuffer = await response.arrayBuffer();
      return Buffer.from(arrayBuffer);
    } catch (error) {
      debug('R2 download error: %o', error);
      throw new Error(`Failed to download from R2: ${error.message}`);
    }
  }

  upload(bucket, key, filepath) {
    // Generate unique key synchronously so it can be returned immediately
    const unique_key = storage.unique_name(key);

    // Read file from disk and upload
    if (fs.existsSync(filepath)) {
      const data = fs.readFileSync(filepath);
      // Call internal version that doesn't generate another unique key
      const uploadPromise = this._upload_stream_with_key(bucket, unique_key, data);
      return [unique_key, uploadPromise];
    }

    debug('[storage.upload] File not found: %s', filepath);
    throw new Error(`upload(): file not found: ${filepath}`);
  }

  async _upload_stream_with_key(bucket, key, data) {
    debug(
      '[storage._upload_stream_with_key] Starting upload: bucket=%s, key=%s, data_size=%d',
      bucket,
      key,
      data.length
    );

    if (!this.r2_enabled) {
      debug('R2 not configured, skipping upload');
      return key;
    }


    debug('[storage._upload_stream_with_key] Outbound handler URL: %s', storage.outbound_url);

    const buffer = this._toBuffer(data);
    debug('[storage._upload_stream_with_key] Uploading key=%s, buffer size: %d', key, buffer.length);

    try {
      const resultKey = await this._upload_bytes(key, buffer);
      debug('[storage._upload_stream_with_key] Upload successful, returned key: %s', resultKey);
      return resultKey;
    } catch (error) {
      debug('R2 upload error: %o', error);
      throw new Error(`Failed to upload to R2: ${error.message}`);
    }
  }

  async download(bucket, key, filepath) {
    const data = await this.download_stream(bucket, key);

    let real_fp = filepath;
    if (!filepath.startsWith('/tmp')) {
      real_fp = path.join('/tmp', path.resolve(filepath));
    }

    // Write data to file
    fs.mkdirSync(path.dirname(real_fp), { recursive: true });
    fs.writeFileSync(real_fp, data);
  }

  async download_directory(bucket, prefix, out_path) {
    if (!this.r2_enabled) {
      debug('R2 not configured, skipping download_directory');
      return;
    }


    // List objects through the Worker outbound handler.
    const listParams = new URLSearchParams({ bucket, prefix });
    const listUrl = `${storage.outbound_url}/r2/list?${listParams}`;
    
    try {
      const response = await fetch(listUrl, {
        method: 'GET',
        headers: { 'Content-Type': 'application/json' },
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`HTTP ${response.status}: ${errorText}`);
      }

      const result = await response.json();
      const objects = result.objects || [];
      
      for (const obj of objects) {
        const file_name = obj.key;
        const path_to_file = path.dirname(file_name);
        fs.mkdirSync(path.join(out_path, path_to_file), { recursive: true });
        await this.download(bucket, file_name, path.join(out_path, file_name));
      }
    } catch (error) {
      debug('R2 download_directory error: %o', error);
      throw new Error(`Failed to download directory from R2: ${error.message}`);
    }
  }

  async downloadDirectory(bucket, prefix, out_path) {
    return this.download_directory(bucket, prefix, out_path);
  }

  uploadStream(bucket, key) {
    // Return [stream, promise, unique_key] to match native wrapper API
    const unique_key = storage.unique_name(key);
    
    const stream = require('stream');
    const passThrough = new stream.PassThrough();
    const chunks = [];
    
    passThrough.on('data', (chunk) => chunks.push(chunk));
    
    const upload = new Promise((resolve, reject) => {
      passThrough.on('end', async () => {
        try {
          const buffer = Buffer.concat(chunks);
          await this._upload_stream_with_key(bucket, unique_key, buffer);
          resolve();
        } catch (err) {
          reject(err);
        }
      });
      passThrough.on('error', reject);
    });
    
    return [passThrough, upload, unique_key];
  }

  async downloadStream(bucket, key) {
    // Return a Promise that resolves to a readable stream
    const data = await this.download_stream(bucket, key);
    const stream = require('stream');
    const readable = new stream.Readable();
    readable.push(data);
    readable.push(null); // Signal end of stream
    return readable;
  }
}

module.exports.storage = storage;
